src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java

```diff
@@ -435,6 +457,26 @@ public boolean isVMAliveOrHealthy() throws Exception {
         return AzureVMManagementServiceDelegate.isVMAliveOrHealthy(this);
     }
 
+    private final Object publicIPAttachLock = new Object();
+
+    public String attachPublicIP() {
+        if (!publicIP.isEmpty()) {
+            return publicIP;
+        }
+        //attachPublicIP will try and provision a public IP or will return if one already exists
+        // because of that we can just wait here until we have a public IP
+        synchronized (publicIPAttachLock) {
+            AzureVMCloud azureVMCloud = (AzureVMCloud) Jenkins.getInstance().getCloud(cloudName);
```

Comment:
You might want to move this inside the try.  It's probably not critical, but deletion of the parent cloud (or change of name, etc.) will leave existing nodes around but getCloud will fail.